examples/audio/lhotse_example.py

```diff
@@ -0,0 +1,235 @@
+#!/usr/bin/env python
+
+"""
+This is a working example of using a new PyTorch dataloading API called DataPipes
+with Lhotse. It's not the final version and probably a lot of things will change.
+
+Lhotse is a speech data representation library that brings various corpora into
+a single format and interfaces well with PyTorch. See: https://github.com/lhotse-speech/lhotse
+
+Prepare the env like this first:
+
+   $ conda create -n torchdata python=3.8
+   $ conda activate torchdata
+   $ conda install numpy
+   $ pip install --pre torch -f https://download.pytorch.org/whl/nightly/cpu/torch_nightly.html
+   $ pip install --user "git+https://github.com/pytorch/data.git"
+
+Optionally, if not in this Lhotse branch already:
+
+   $ pip install git+https://github.com/lhotse-speech/lhotse@feature/datapipes-prototyping
+
+"""
+
+
+import warnings
+from collections import deque, defaultdict
+from functools import partial
+from pathlib import Path
+from typing import Optional
+
+from lhotse import CutSet, load_manifest
+from lhotse.utils import Seconds
+from lhotse.recipes import download_librispeech, prepare_librispeech
+
+import torchdata
+import torch
+import torchdata.datapipes.iter as dp  # Only use Iter-style DataPipe to illustrate the pipeline
+
+from torch.utils.data.datapipes.iter import Multiplexer, Demultiplexer
+from torch.utils.data.communication.eventloop import SpawnProcessForDataPipeline
+from torch.utils.data.communication.protocol import IterDataPipeQueueProtocolClient
+from torch.utils.data.communication.iter import QueueWrapper
+
+
+class CutsReader(dp.IterDataPipe):
+    def __init__(self, path) -> None:
+        self.cuts = CutSet.from_jsonl_lazy(path)
+
+    def __iter__(self):
+        yield from self.cuts
+
+
+class DurationBatcher(dp.IterDataPipe):
+    def __init__(
+        self,
+        datapipe,
+        max_frames: int = None,
+        max_samples: int = None,
+        max_duration: Seconds = None,
+        max_cuts: Optional[int] = None,
+        drop_last: bool = False,
+    ):
+        from lhotse.dataset.sampling.base import SamplingDiagnostics, TimeConstraint
+
+        self.datapipe = datapipe
+        self.reuse_cuts_buffer = deque()
+        self.drop_last = drop_last
+        self.max_cuts = max_cuts
+        self.diagnostics = SamplingDiagnostics()
+        self.time_constraint = TimeConstraint(
+            max_duration=max_duration, max_frames=max_frames, max_samples=max_samples
+        )
+
+    def __iter__(self):
+        self.datapipe = iter(self.datapipe)
```

Comment:
Please do not remove the reference to prior/source datapipe here. Since we want to let `DataLoaderV2` have `graph` mode execution of data pipeline (run parallel, automatic sharding, etc.), we need to keep reference graph of each DataPipe instance.

Reply:
This would also prevent you do second iteration.

Reply:
Thanks. DataLoaderV2 sounds interesting -- is there a description somewhere that I could read to learn more?